backend/src/main/java/com/javahelp/backend/domain/user/register/UserRegisterInteractor.java

```diff
@@ -0,0 +1,126 @@
+package com.javahelp.backend.domain.user.register;
+
+import com.javahelp.backend.data.IUserStore;
+import com.javahelp.model.user.User;
+import com.javahelp.model.user.UserInfo;
+import com.javahelp.model.user.UserPassword;
+
+/**
+ * An interactor to make new {@link User} and {@link UserPassword} for all types of users
+ */
+public abstract class UserRegisterInteractor <T extends IUserRegisterInputBoundary,
+        U extends UserInfo>{
+
+    IUserStore userStore;
+
+    /**
+     * Creates a new {@link UserRegisterInteractor} instance
+     *
+     * @param userStore  {@link IUserStore} to use
+     */
+    public UserRegisterInteractor(IUserStore userStore) {
+        this.userStore = userStore;
+    }
+
+    /**
+     * Generate a new pseudo {@link User}. Private because this is exclusively for a general
+     * {@link User} type.
+     *
+     * @return a new {@link User} entity
+     */
+    private User register(T boundary){
+        return new User("", this.bundleInfo(boundary), boundary.getUsername());
+    }
+
+    /**
+     * Checking for violations for registering a {@link User} account.
+     * @param boundary an instance of {@link IUserRegisterInputBoundary} to retrieve input info
+     *
+     * @return a {@link RegisterUserResponse} model that contains message of failure at registration
+     * Null if all tests passed.
+     */
+    protected RegisterUserResponse isValidRegistering(T boundary){
+        if ((IUserStore.getDefaultImplementation().readByEmail(boundary.getEmailAddress()) != null)
+                && (IUserStore.getDefaultImplementation().readByUsername(boundary.getUsername()) != null)){
+            // If both username and email address exist, then return both errors
+            return new RegisterUserResponse(null, null, "Both the given email " +
+                    "address and username are already occupied. If that is previously yours, " +
+                    "consider logging in instead?");
+        }
+        if(IUserStore.getDefaultImplementation().readByEmail(boundary.getEmailAddress()) != null){
+            return new RegisterUserResponse(null, null, "The email" +
+                    " address is already occupied. If that is previously yours, consider logging " +
+                    "in instead?");
+        }
+        if(IUserStore.getDefaultImplementation().readByUsername(boundary.getUsername()) != null){
+            return new RegisterUserResponse(null, null, "The username" +
+                    " is already occupied. If that is previously yours, consider logging in instead?");
```

Comment:
These error messages should communicate what went wrong, including suggestions like 'if that is previously yours consider logging in instead' is probably not what we want to do. IMHO its better to have a simple error message like 'Username already taken' and if the frontend wants to show a message to the end user, it can figure that out itself.

Reply:
Sure